[Product Creation with AI: v2] Logic to switch options and undo edits

selanthiraiyan opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #13110 
<!-- Id number of the GitHub issue this PR addresses. -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

Adds a way to switch between AI suggested name, description and short description options. Adds logic for the "Undo edits" functionality. 

Changes
- Logic to show/hide switch option view based on the available number of AI suggested options.
    - The switch option buttons should be displayed only if there is more than one option for name, description fields.
    - The switch option buttons should enable or disable based on the number of options and the currently viewed options.
- Logic to save and undo edits.
    - The "Undo edits" button should appear when any of the name, description and short description value is edited.
    - The edited value should be retained when switching between options.
    - The edited value should be reverted to the original value received from remote upon tapping "Undo edits" button.
- Update init method to save product with edited name, description and short description values. 
- Unit tests.

## Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->

1. Create a JN site with Jetpack AI plugin
2. Turn on `productCreationAIv2M1` feature flag
3. Build and run the app
4. Navigate to the products tab
5. Tap `+` button on top right
6. Tap "Create a product with AI"
7.  Select a package photo or enter product features.
8. Tap "Generate Product Details", and you will land on the new product preview screen
9. Try switching between options by tapping the arrows in the "Name, Summary & Description" section
10. Try editing any of the text fields, and you should see an option to "Undo edits"
11. Tapping on "Undo edits" should revert the changes. 

## Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->
- Test the undo edits works as expected on all text fields. (Name, Short description and Description fields)
- Test that you can save the product with an edited value for the Name, Short description and Description fields. 
- Test that switching between options works as expected.

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->

#### Switch between options
| Only one option | Three options | No options |
|--------|--------|--------|
| ![NoOptions](https://github.com/woocommerce/woocommerce-ios/assets/524475/4f04f13a-f21e-4edf-9fa7-721711704c08) | ![image](https://github.com/woocommerce/woocommerce-ios/assets/524475/5b713485-620a-4d69-944f-4b94f5bb3d0e) | ![Simulator Screenshot - iPhone 15 Pro Max - 2024-07-09 at 18 55 33](https://github.com/woocommerce/woocommerce-ios/assets/524475/3de7a9be-3add-4849-a2b7-77de37eb99b5) | 

#### Undo edits
| Name | Short description | Description |
|--------|--------|--------|
| ![image](https://github.com/woocommerce/woocommerce-ios/assets/524475/30a08a8d-738f-4298-a6c2-ae85543ef7fc) | ![image](https://github.com/woocommerce/woocommerce-ios/assets/524475/e31a765f-2a24-492e-ab3f-ac9bf483127b) | ![image](https://github.com/woocommerce/woocommerce-ios/assets/524475/774208fc-12f6-488b-b5c4-397821478c75) | 

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

selanthiraiyan committed:
Merge branch 'feat/13265-short-description' into feat/13110-switch-undo-logic

selanthiraiyan committed:
Merge branch 'trunk' into feat/13110-switch-undo-logic

selanthiraiyan committed:
Get name and summary info to initialize a product from AIProduct.

selanthiraiyan committed:
Pass names and description values from legacy screen.

selanthiraiyan committed:
Move Field enum to view model.

selanthiraiyan committed:
Enum to hold name and summary options.

selanthiraiyan committed:
Add properties to store selected option.

selanthiraiyan committed:
Helper to display product information.

selanthiraiyan committed:
String for options title.

selanthiraiyan committed:
Return AIProduct from helper method.

selanthiraiyan committed:
Add helper method to turn AIProduct into Product.

selanthiraiyan committed:
Store AI product instead of Product.

selanthiraiyan committed:
Error when no options found in AI product.

selanthiraiyan committed:
Property to show options switch title.

selanthiraiyan committed:
Properties to control switching between suggestions.

selanthiraiyan committed:
Helper methods to switch between options.

selanthiraiyan committed:
Helper method to undo edits.

selanthiraiyan committed:
Observe selected option to update currently selected option on screen.

selanthiraiyan committed:
Update logic to decide when to display undo edits view.

selanthiraiyan committed:
Show options title.

selanthiraiyan committed:
Call action to Switch between options.

selanthiraiyan committed:
Call action to undo edits.

selanthiraiyan committed:
Update init method to pass name and summary info.

selanthiraiyan committed:
Disable switch options buttons when applicable.

selanthiraiyan committed:
Hide switch options view when applicable.

selanthiraiyan committed:
Update unit tests based on new changes.

selanthiraiyan committed:
Unit test that switching between options works as expected.

selanthiraiyan committed:
Update unit test to use different shipping values.

selanthiraiyan committed:
Merge branch 'trunk' into feat/13110-switch-undo-logic

dangermattic commented:
<!--
  0 Errors
  2 Warnings: This PR is larger than 500 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 500 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/213">19.5</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

itsmeichigo commented:
Hi Sharma, I was just testing the feature and found that tapping on Undo edits of a field would revert all changes on all 3 fields. Is this intentional?


https://github.com/woocommerce/woocommerce-ios/assets/5533851/ceced565-fd7f-463f-af27-7d10f4323b8e

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Products/Add Product/AddProductWithAI/Preview/ProductDetailPreviewView.swift:
```diff
@@ -169,10 +169,9 @@ private extension ProductDetailPreviewView {
                           onUndoEdits: {
-            // TODO: update this
-            viewModel.productName = viewModel.generatedProduct?.name ?? ""
+            viewModel.undoEdits(in: ProductDetailPreviewViewModel.ProductDetailField.name)
```
Nit: this can be simplified:

```suggestion
            viewModel.undoEdits(in: .name)
```

selanthiraiyan replied:
Done in fb9019c11a9c0fada795e74bb5db2a1f7bf8cf33

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Products/Add Product/AddProductWithAI/Preview/ProductDetailPreviewView.swift:
```diff
@@ -169,10 +169,9 @@ private extension ProductDetailPreviewView {
+            viewModel.undoEdits(in: ProductDetailPreviewViewModel.ProductDetailField.name)
         })
-        .focused($focusedField, equals: FocusedField.name)
+        .focused($focusedField, equals: ProductDetailPreviewViewModel.ProductDetailField.name)
```
Nit:

```suggestion
        .focused($focusedField, equals: .name)
```

Similar updates can be made in the other places below for brevity.

selanthiraiyan replied:
Done in fb9019c11a9c0fada795e74bb5db2a1f7bf8cf33

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Products/Add Product/AddProductWithAI/Preview/ProductDetailPreviewViewModel.swift:
```diff
@@ -7,12 +7,29 @@ import protocol Storage.StorageManagerType
 /// View model for `ProductDetailPreviewView`
 ///
 final class ProductDetailPreviewViewModel: ObservableObject {
+    enum ProductDetailField: Equatable {
```
Super nit: this could be renamed to `EditableField` for clarity. There's no need to repeat the `ProductDetail` prefix.

selanthiraiyan replied:
Done in 9330760bd52d542cc68166a849e665e809dc632c

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Products/Add Product/AddProductWithAI/Preview/ProductDetailPreviewViewModel.swift:
```diff
@@ -7,12 +7,29 @@ import protocol Storage.StorageManagerType
-    @Published private(set) var generatedProduct: Product?
+    @Published private(set) var generatedAIProduct: AIProduct?
+
+    @Published private(set) var selectedOptionIndex: Int = 0
```
Super nit:

```suggestion
    @Published private(set) var selectedOptionIndex = 0
```

selanthiraiyan replied:
Done in 2b65f09d5fd0c101c3519b8667bbcaf1ad4475e6

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Products/Add Product/AddProductWithAI/Preview/ProductDetailPreviewViewModel.swift:
```diff
@@ -30,28 +47,51 @@ final class ProductDetailPreviewViewModel: ObservableObject {
+
     var hasChangesToProductName: Bool {
-        guard let generatedProduct else {
+        guard let generatedAIProduct else {
```
Should avoid crashes by ensuring we can subscript the names array as well?

```suggestion
        guard let generatedAIProduct, generatedAIProduct.names.count > selectedOptionIndex  else {
```

Similar comment for the other fields.

selanthiraiyan replied:
I started reading values from the array using `safe:` in ebc6bcd862f40711308fb3c76045f1757384d620

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Products/Add Product/AddProductWithAI/Preview/ProductDetailPreviewViewModel.swift:
```diff
@@ -170,6 +213,29 @@ final class ProductDetailPreviewViewModel: ObservableObject {
     }
 
+    // MARK: Switch options
+    func next() {
```
Nit: this method name could be improved for clarity:

```suggestion
    func switchToNextOption() {
```

selanthiraiyan replied:
Done in 163c44fb3ac21a2db3147551b850e0d4de2221a7

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Products/Add Product/AddProductWithAI/Preview/ProductDetailPreviewViewModel.swift:
```diff
@@ -170,6 +213,29 @@ final class ProductDetailPreviewViewModel: ObservableObject {
+        selectedOptionIndex = selectedOptionIndex + 1
+    }
+
+    func previous() {
```
Similar comment as above:

```suggestion
    func switchToPreviousOption() {
```

selanthiraiyan replied:
Done in 163c44fb3ac21a2db3147551b850e0d4de2221a7

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Products/Add Product/AddProductWithAI/Preview/ProductDetailPreviewViewModel.swift:
```diff
@@ -187,22 +253,81 @@ final class ProductDetailPreviewViewModel: ObservableObject {
+            options[selectedOptionIndex] = NameSummaryDescOption(name: option.name,
+                                                                 shortDescription: option.shortDescription,
+                                                                 description: generatedAIProduct.descriptions[selectedOptionIndex])
+        }
```
What do you think about avoiding the duplication:

```suggestion
        let name = updatedField != .name ? productName :
            generatedAIProduct.names[selectedOptionIndex]

        let shortDescription = updatedField != .shortDescription ? productShortDescription :
            generatedAIProduct.shortDescriptions[selectedOptionIndex]

        let description = updatedField != .description ? productDescription :
            generatedAIProduct.descriptions[selectedOptionIndex]

        options[selectedOptionIndex] = NameSummaryDescOption(name: name, shortDescription: shortDescription, description: description)
```

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Products/Add Product/AddProductWithAI/Preview/ProductDetailPreviewViewModel.swift:
```diff
@@ -187,22 +253,81 @@ final class ProductDetailPreviewViewModel: ObservableObject {
-        productDescription = product.fullDescription ?? ""
+    func displayAIProductDetails(aiProduct: AIProduct) throws {
+        guard
+            aiProduct.names.count > 0,
+            aiProduct.shortDescriptions.count > 0,
+            aiProduct.descriptions.count > 0 else {
```
Nit: We can use `isNotEmpty` for brevity here.

selanthiraiyan replied:
Done in f4567a75133aedb8883e110eb145fc070a9e61d5

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Products/Add Product/AddProductWithAI/Preview/ProductDetailPreviewViewModel.swift:
```diff
@@ -7,12 +7,29 @@ import protocol Storage.StorageManagerType
+    @Published private(set) var generatedAIProduct: AIProduct?
+
+    @Published private(set) var selectedOptionIndex: Int = 0
+    @Published private var options: [NameSummaryDescOption] = [NameSummaryDescOption(name: "",
+                                                                                     shortDescription: "",
+                                                                                     description: "")]
```
❓ I'm curious why we need an empty object for the default value instead of an empty array?

selanthiraiyan replied:
Good point. An empty array is enough. Done in 580dd7448838c2334833c93a6a30e6883d4926c0

selanthiraiyan committed:
Undo only specific field and for other fields use the texts from textfields while initiating option.

selanthiraiyan committed:
Test other fields are not reverted while undoing.

selanthiraiyan commented:
> I was just testing the feature and found that tapping on Undo edits of a field would revert all changes on all 3 fields. Is this intentional?

Thanks, @itsmeichigo! Sorry, this is a bug that I missed to catch while refactoring. I addressed it in https://github.com/woocommerce/woocommerce-ios/pull/13281/commits/b7110ca567510b8c742d105b399841d78b56aa46 and added unit tests in https://github.com/woocommerce/woocommerce-ios/pull/13281/commits/b827514ca66266892aaa81a7b262a3061010c1b6.

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Products/Add Product/AddProductWithAI/Preview/ProductDetailPreviewViewModel.swift:
```diff
@@ -187,22 +253,81 @@ final class ProductDetailPreviewViewModel: ObservableObject {
+            return
+        }
+
+        switch field {
+        case .name:
+            options[selectedOptionIndex] = NameSummaryDescOption(name: generatedAIProduct.names[selectedOptionIndex],
+                                                                 shortDescription: productShortDescription,
+                                                                 description: productDescription)
+        case .shortDescription:
+
+            options[selectedOptionIndex] = NameSummaryDescOption(name: productName,
+                                                                 shortDescription: generatedAIProduct.shortDescriptions[selectedOptionIndex],
+                                                                 description: productDescription)
+        case .description:
+            options[selectedOptionIndex] = NameSummaryDescOption(name: productName,
+                                                                 shortDescription: productShortDescription,
+                                                                 description: generatedAIProduct.descriptions[selectedOptionIndex])
+        }
```
WDYT about avoiding the duplicated logic:

```suggestion
        let name = updatedField != .name ? productName :
            generatedAIProduct.names[selectedOptionIndex]

        let shortDescription = updatedField != .shortDescription ? productShortDescription :
            generatedAIProduct.shortDescriptions[selectedOptionIndex]

        let description = updatedField != .description ? productDescription :
            generatedAIProduct.descriptions[selectedOptionIndex]

        options[selectedOptionIndex] = NameSummaryDescOption(name: name, shortDescription: shortDescription, description: description)
```

selanthiraiyan replied:
Nice suggestion. Added in https://github.com/woocommerce/woocommerce-ios/pull/13281/commits/8dc33ca19785c0cdbf2426296fd30d0e3c505cfb

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Products/Add Product/AddProductWithAI/Preview/ProductDetailPreviewViewModel.swift:
```diff
@@ -187,22 +253,81 @@ final class ProductDetailPreviewViewModel: ObservableObject {
+            return
+        }
+
+        guard let option = options[safe: selectedOptionIndex] else {
+            return
+        }
```
I think `option` is unused here.

selanthiraiyan replied:
Good catch. Removed in d2a51b7a25e5e11c4936f4dacc3289eef1423aad.

itsmeichigo reviewed (approved):
Thanks for the updates Sharma! This works as expected now 👍 

I left a few nit-pickings in the comments. I'm also unsure if I understand the changes around the saving logic, it would be nice if you could share the rationale behind that (maybe in the PR description next time too).

selanthiraiyan committed:
Rename textfield enum.

selanthiraiyan committed:
Simplify field value in method calls.

selanthiraiyan committed:
Remove unused option value.

selanthiraiyan committed:
Assign empty array as initial value for options.

selanthiraiyan committed:
Remove explicit type.

selanthiraiyan committed:
Safely read values using selectedOptionIndex.

selanthiraiyan committed:
Rename methods for clarity.

selanthiraiyan committed:
Use isNotEmpty for brevity.

selanthiraiyan committed:
Refactor to remove duplicated code.

selanthiraiyan committed:
Merge branch 'trunk' into feat/13110-switch-undo-logic

selanthiraiyan commented:
Thank you for the review and nice suggestions, @itsmeichigo! I have pushed the necessary updated. 

> I'm also unsure if I understand the changes around the saving logic, it would be nice if you could share the rationale behind that (maybe in the PR description next time too).

Sorry, my bad. I updated the PR description to add more information under the "Changes" title. Kindly let me know if that helps. 

Please take another look at the PR when you can. 🙇

itsmeichigo reviewed (approved):
Thanks for the updates Sharma :shipit:
